File: src/dataset/cpp/nucleotide-count_exercism-pawelo091991/openAI/ChatGPT4_nucleotide_count_v2.cpp
#include "nucleotide_count.h"
#include <unordered_map>

namespace dna {
	counter::counter(const std::string& input) : data{{'A', 0}, {'T', 0}, {'C', 0}, {'G', 0}} {
		for (char nucleotide : input) {
			if (data.find(nucleotide) != data.end()) {
				data[nucleotide]++;
			} else {
				throw std::invalid_argument("invalid nucleotide");
			}
		}
	}

	std::unordered_map<char, int> counter::nucleotide_counts() const {
		return data;
	}

	int counter::count(char input) const {
		auto it = data.find(input);
		if (it != data.end()) {
			return it->second;
		} else {
			throw std::invalid_argument("invalid input argument");
		}
	}
}

File: src/dataset/typescript/nucleotide-count_exercism-wackerow/openAI/ChatGPT4_nucleotide-count_v2.ts
type Count = {
  A: number,
  C: number,
  T: number,
  G: number,
}

class NucleotideCount {
  static nucleotideCounts(dna: string): Count {
    const counts: Count = { A: 0, C: 0, T: 0, G: 0 };

    for (const nucleotide of dna.toUpperCase()) {
      if (!(nucleotide in counts)) {
        throw 'Invalid nucleotide in strand';
      }
      counts[nucleotide as keyof Count]++;
    }

    return counts;
  }
}

export default NucleotideCount;